aztec/src/main/kotlin/org/wordpress/aztec/watchers/event/sequence/ObservationQueue.kt

```diff
@@ -0,0 +1,65 @@
+package org.wordpress.aztec.watchers.event.sequence
+
+import android.os.Build
+import org.wordpress.aztec.watchers.event.IEventInjector
+import org.wordpress.aztec.watchers.event.buckets.API26Bucket
+import org.wordpress.aztec.watchers.event.text.TextWatcherEvent
+import org.wordpress.aztec.watchers.event.buckets.Bucket
+
+class ObservationQueue(injector: IEventInjector) : EventSequence<TextWatcherEvent>() {
+    val buckets = ArrayList<Bucket>()
+    val injector = injector
+
+    init {
+        init()
+    }
+
+    override fun add(element: TextWatcherEvent): Boolean {
+        synchronized(this@ObservationQueue) {
+            val added: Boolean = super.add(element)
+            if (added) {
+                processQueue()
+            }
+            return added
+        }
+    }
+
+    private fun processQueue() {
+        // here let's check whether our current queue matches / fits any of the installed buckets
+        var foundOnePartialMatch = false
+        for (bucket in buckets) {
+            for (operation in bucket.userOperations) {
+                if (size < operation.sequence.size) {
+                    if (operation.isUserOperationPartiallyObservedInSequence(this)) {
+                        foundOnePartialMatch = true
+                    }
+                } else {
+                    // does this particular event look like a part of any of the user operations as defined in this bucket?
+                    if (operation.isUserOperationObservedInSequence(this)) {
+                        // replace user operation with ONE TextWatcherEvent and inject this one in the actual
+                        // textwatchers
+                        val replacementEvent = operation.buildReplacementEventWithSequenceData(this)
+                        injector.executeEvent(replacementEvent)
+                        clear()
+                    }
+                }
+            }
+        }
+
+        // we didn't find neither a partial match nor a total match, let's just clear the queue
+        if (size > 0 && !foundOnePartialMatch) {
+            // immediately discard the queue
+            clear()
+        }
+    }
+
+    private fun init() {
```

Comment:
Wondering if we can remove this method and move the code  to default `init` above.

Reply:
Good call, I had plans for `init()` to be called in some other places but in the end it wasn't necessary. Modified this in c7f89b1